llvm/test/TableGen/ReturnRange.td

```diff
@@ -0,0 +1,108 @@
+  let Fields = [
+     "Name", "Encoding", "FeaturesRequired",
+  ];
+
+  let PrimaryKey = [ "Encoding" ];
+  let PrimaryKeyName = "lookupSysRegByEncoding";
+  let PrimaryKeyReturnRange = true;
+}
+
+let FeaturesRequired = [{ {Feature1} }] in {
+def : SysReg<"csr1", 0x7C0>;
+}
+
+let FeaturesRequired = [{ {Feature2} }] in {
+def : SysReg<"csr2", 0x7C0>;
+}
+
+// CHECK: #ifdef GET_List1_DECL
+// CHECK-NEXT: llvm::iterator_range<const SysReg *> lookupSysRegByEncoding(uint16_t Encoding);
+// CHECK-NEXT: #endif
+
+// CHECK: #ifdef GET_List1_IMPL
+// CHECK-NEXT: constexpr SysReg List1[] = {
+// CHECK-NEXT:   { "csr1", 0x7C0,  {Feature1}  }, // 0
+// CHECK-NEXT:   { "csr2", 0x7C0,  {Feature2}  }, // 1
+// CHECK-NEXT:  };
+
+// CHECK: llvm::iterator_range<const SysReg *> lookupSysRegByEncoding(uint16_t Encoding) {
+// CHECK-NEXT:   SysReg Key;
+// CHECK-NEXT:   Key.Encoding = Encoding;
+// CHECK-NEXT:   auto Table = ArrayRef(List1);
+// CHECK-NEXT:   auto It = std::equal_range(Table.begin(), Table.end(), Key,
+// CHECK-NEXT:     [](const SysReg &LHS, const SysReg &RHS) {
+// CHECK-NEXT:       if (LHS.Encoding < RHS.Encoding)
+// CHECK-NEXT:         return true;
+// CHECK-NEXT:       if (LHS.Encoding > RHS.Encoding)
+// CHECK-NEXT:         return false;
+// CHECK-NEXT:       return false;
+// CHECK-NEXT:     });
+
+// CHECK:   return llvm::make_range(It.first, It.second);
+// CHECK-NEXT: }
+// CHECK-NEXT: #endif
+
+def List2 : GenericTable {
+  let FilterClass = "SysReg";
+  let Fields = [
+     "Name", "Encoding", "FeaturesRequired",
+  ];
+}
+
+def lookupSysRegByName : SearchIndex {
+  let Table = List2;
+  let Key = [ "Name" ];
+  let ReturnRange = true;
+}
+
+// CHECK: #ifdef GET_List2_DECL
+// CHECK-NEXT: llvm::iterator_range<const SysReg *> lookupSysRegByName(StringRef Name);
+// CHECK-NEXT: #endif
+
+// CHECK: #ifdef GET_List2_IMPL
+// CHECK-NEXT: constexpr SysReg List2[] = {
+// CHECK-NEXT:   { "csr1", 0x7C0,  {Feature1}  }, // 0
+// CHECK-NEXT:   { "csr2", 0x7C0,  {Feature2}  }, // 1
+// CHECK-NEXT:  };
+
+// CHECK: llvm::iterator_range<const SysReg *> lookupSysRegByName(StringRef Name) {
+// CHECK-NEXT:   struct IndexType {
+// CHECK-NEXT:     const char * Name;
+// CHECK-NEXT:     unsigned _index;
+// CHECK-NEXT:   };
+// CHECK-NEXT:   static const struct IndexType Index[] = {
+// CHECK-NEXT:     { "CSR1", 0 },
+// CHECK-NEXT:     { "CSR2", 1 },
+// CHECK-NEXT:   };
+
+// CHECK:   IndexType Key;
+// CHECK-NEXT:   Key.Name = Name.upper();
+// CHECK-NEXT:   auto Table = ArrayRef(Index);
+// CHECK-NEXT:   auto It = std::equal_range(Table.begin(), Table.end(), Key,
+// CHECK-NEXT:     [](const IndexType &LHS, const IndexType &RHS) {
+// CHECK-NEXT:       int CmpName = StringRef(LHS.Name).compare(RHS.Name);
+// CHECK-NEXT:       if (CmpName < 0) return true;
+// CHECK-NEXT:       if (CmpName > 0) return false;
+// CHECK-NEXT:       return false;
+// CHECK-NEXT:     });
+
+// CHECK:   return llvm::make_range(It.first, It.second);
```

Comment:
The iterator here is `IndexType *` right? That's not convertible to `llvm::iterator_range<const SysReg *>` is it?

Reply:
Thanks for pointing this out!. I have fixed this in the latest patchset. This required emitting a separate global table sorted by the field which we are trying to lookup with (for example, in the test it is `Name`), so that a range of entries can be returned. Do let me know what you think of it. 

Though, there is currently no need of emitting a different signature for non-primary functions both upstream and in downstream I think, therefore if it is felt that there is so much of redundant code-changes, then at this point maybe we can just support emitting different signature for primary functions only.

Reply:
I think we can drop the support for ReturnRange and !PrimaryKey for now.

Reply:
Just to be clear, I should only allow emitting different signature from primary lookupfunctions and not support emitting different signature for functions other than primary lookupfunctions (eg. `lookupsysregbyname`). 
For primary function , there is already a use case of encoding conflict, so would need to emit a different signature. However, there is no use case for emitting a different signature for non-primary functions(eg. `lookupsysregbyname`), so probably the added complexity is not worth it for the time being.

Apologies if I misunderstood your comment or missed something. This comment is just to clarify on what needs to  be dropped and what can be retained.

Reply:
`Just to be clear, I should only allow emitting different signature from primary lookupfunctions and not support emitting different signature for functions other than primary lookupfunctions`

Yes.

Sorry I meant to write `IsPrimary` instead of `PrimaryKey` in my reply. I fixed it.